Src/DependencyCollector/Microsoft.AI.DependencyCollector/DependencyCollectorDiagnosticListener.cs

```diff
@@ -0,0 +1,235 @@
+                    }
+
+                    // Add the parent ID
+                    string parentId = telemetry.Id;
+                    if (!string.IsNullOrEmpty(parentId) && !requestHeaders.Contains(RequestResponseHeaders.StandardParentIdHeader))
+                    {
+                        requestHeaders.Add(RequestResponseHeaders.StandardParentIdHeader, parentId);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Diagnostic event handler method for 'System.Net.Http.Response' event. Even in the case of an exception, this will still be called.
+        /// See https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L71 for more information.
+        /// </summary>
+        [DiagnosticName("System.Net.Http.Response")]
+        public void OnResponse(HttpResponseMessage response, Guid loggingRequestId)
+        {
+            if (response != null)
+            {
+                DependencyTelemetry telemetry;
+                if (this.pendingTelemetry.TryRemove(loggingRequestId, out telemetry))
+                {
+                    string targetApplicationId = GetRequestContextKeyValue(response.Headers, RequestResponseHeaders.RequestContextTargetKey);
+                    if (!string.IsNullOrEmpty(targetApplicationId) && !string.IsNullOrEmpty(telemetry.Context.InstrumentationKey))
+                    {
+                        // We only add the cross component correlation key if the key does not represent the current component.
+                        string sourceApplicationId;
+                        if (this.correlationIdLookupHelper.TryGetXComponentCorrelationId(telemetry.Context.InstrumentationKey, out sourceApplicationId) &&
+                            targetApplicationId != sourceApplicationId)
+                        {
+                            telemetry.Type = RemoteDependencyConstants.AI;
+                            telemetry.Target += " | " + targetApplicationId;
+                        }
+                    }
+
+                    int statusCode = (int)response.StatusCode;
+                    telemetry.ResultCode = (0 < statusCode) ? statusCode.ToString(CultureInfo.InvariantCulture) : string.Empty;
+                    telemetry.Success = (0 < statusCode) && (statusCode < 400);
+
+                    telemetry.Stop();
+                    this.client.Track(telemetry);
+                }
+            }
+        }
+
+        internal static IEnumerable<string> GetHeaderValues(HttpHeaders headers, string headerName)
+        {
+            IEnumerable<string> result;
+            if (headers == null || !headers.TryGetValues(headerName, out result))
+            {
+                result = Enumerable.Empty<string>();
+            }
+            return result;
+        }
+
+        internal static string GetHeaderKeyValue(HttpHeaders headers, string headerName, string keyName)
```

Comment:
My concern is DependencyCollectorDiagnosticsListener.GetHeaderKeyValue doesn't make any sense. It is a good thing to strive to test everything but that shouldn't result in classes that are harder to make sense of. So for example, if they are only used by this class one way to test this functionality would be to test these through the classes internal / public methods. If you tests are telling you to make something internal when they aren't truly internal - that's a good indication that there is less cohesion in the class - that's where my suggestion about fixing it by moving it to a utility class comes.